roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx

```diff
@@ -0,0 +1,325 @@
+                                                       GlobalObservables global_observables, std::string global_observables_tag) {
+   auto sim_pdf = dynamic_cast<RooSimultaneous *>(pdf);
+   if (sim_pdf == nullptr) {
+      throw std::logic_error("Can only build RooSumL from RooSimultaneous pdf!");
+   }
+
+   // the rest of this function is an adaptation of RooAbsTestStatistic::initSimMode:
+
+   RooAbsCategoryLValue &simCat = (RooAbsCategoryLValue &)sim_pdf->indexCat();
+
+   // FIXME: process_empty_data_sets was previously member function processEmptyDataSets(), which returned extended_ for RooSumL (previously named RooSimultaneousL). See RooAbsL and RooSumL/RooUnbinnedL implementations. I think this should work here, but be careful in other subclasses!
+   bool process_empty_data_sets = RooAbsL::isExtendedHelper(pdf, extended);
+
+   TString simCatName(simCat.GetName());
+   // Note: important not to use cloned dataset here (possible when this code is run in Roo[...]L ctor), use the
+   // original one (which is data_ in Roo[...]L ctors, but data here)
+   std::unique_ptr<TList> dsetList{data->split(simCat, process_empty_data_sets)};
+   if (!dsetList) {
+      throw std::logic_error("buildSimultaneousLikelihood ERROR, index category of simultaneous pdf is missing in dataset, aborting");
+   }
+
+   // Count number of used states
+   std::size_t N_components = 0;
+
+   RooCatType *type;
+   std::unique_ptr<TIterator> catIter{simCat.typeIterator()};
+   while ((type = (RooCatType *)catIter->Next())) {
+      // Retrieve the PDF for this simCat state
+      RooAbsPdf *component_pdf = sim_pdf->getPdf(type->GetName());
+      auto dset = (RooAbsData *)dsetList->FindObject(type->GetName());
+
+      if (component_pdf && dset && (0. != dset->sumEntries() || process_empty_data_sets)) {
+         ++N_components;
+      }
+   }
+
+   // Allocate arrays
+   std::vector<std::unique_ptr<RooAbsL>> components;
+   components.reserve(N_components);
+   //   _gofSplitMode.resize(N_components);  // not used, Hybrid mode only, see below
+
+   // Create array of regular fit contexts, containing subset of data and single fitCat PDF
+   catIter->Reset();
+   std::size_t n = 0;
+   while ((type = (RooCatType *)catIter->Next())) {
+      // Retrieve the PDF for this simCat state
+      RooAbsPdf *component_pdf = sim_pdf->getPdf(type->GetName());
+      auto dset = (RooAbsData *)dsetList->FindObject(type->GetName());
+
+      if (component_pdf && dset && (0. != dset->sumEntries() || process_empty_data_sets)) {
+         ooccoutI((TObject *)nullptr, Fitting)
+         << "RooSumL: creating slave calculator #" << n << " for state " << type->GetName() << " ("
+                                                   << dset->numEntries() << " dataset entries)" << std::endl;
+
+         // *** START HERE
+         // WVE HACK determine if we have a RooRealSumPdf and then treat it like a binned likelihood
```

Comment:
Can we do without hack?

Reply:
Yeah, I'll factor this out into a separate function, which is necessary for the single-entry builder function that we still intended to build (https://github.com/root-project/root/pull/8700#issue-692681999 todo item 2).

Reply:
This has been resolved in cc3e6be.